### PR TITLE
Add include_terminal param

### LINKED ISSUE #783: Extra param in edge_diffs to output all the final edges
I use the `edge_diffs()` iterator quite a lot. Occasionally I would like to do something with all the edges that remain at the end of the iteration. There might be a workaround, but the easiest thing to me would be if we could call something like `edge_diffs(include_terminal_edges=True)` which would report one extra iteration, listing the final `edges_out` that would be removed to clear the edges at the end of the sequence. Any objections to this?

### PR DESCRIPTION
**Pull request by hyanwong**
Fixes #783

### CODE REVIEW

File: c/tskit/trees.c
Comment (jeromekelleher) [inline c/tskit/trees.c:4375]: Let's make this a ``tsk_flags_t options`` argument - this is the approach taken elsewhere in the API, as it's more flexible and breaks less code when new options are added.

File: python/tests/test_highlevel.py
Comment (jeromekelleher) [inline python/tests/test_highlevel.py:662]: We need to test the value of left and right here as well. What are they?

File: python/tskit/trees.py
Comment (jeromekelleher) [inline python/tskit/trees.py:3465]: What's be value of left and right here?

File: c/tskit/trees.c
Comment (jeromekelleher) [inline c/tskit/trees.c:4511]: We need some C tests of this to make sure we're not touching any uninitialised memory. Easiest thing to do is to run through the iterator setting a ``parent`` array as usual, and then check at the end that the parent array is all -1 if TSK_INCLUDE_TERMINAL has been set.

**Review by jeromekelleher — CHANGES_REQUESTED** [review]
Looks good, but needs more tests. Definitely on the C side we want to exercise this on a few different topology types - what happens on an empty ts? An ordinary ts with several trees? We have to think through all the corner cases.

File: python/tskit/trees.py
Comment (hyanwong) [reply python/tskit/trees.py:3465]: Good point. By default it is `right==left==sequence_length`, which seems reasonable to me, so I've documented this and added to to the high_level tests.

File: c/tskit/trees.c
Comment (hyanwong) [reply c/tskit/trees.c:4511]: Just done this in https://github.com/tskit-dev/tskit/pull/787/commits/b4a3f1d9d26c44b8330bb9bf5276ee7e7c621093 - I've not written many C tests before, so let me know if anything is wrong / unconventional.

**Comment by hyanwong** [comment]
> Looks good, but needs more tests. Definitely on the C side we want to exercise this on a few different topology types - what happens on an empty ts? An ordinary ts with several trees? We have to think through all the corner cases.

Can we not test an empty ts, multi-tree ts etc at the python level? We have a function `get_example_tree_sequences()` in test_highlevel.py which can return various pathological tree sequences like this. Actually, having looked at it, I see it doesn't test a ts with no edges, so I'll open a PR for that.

**Comment by jeromekelleher** [comment]
> Can we not test an empty ts, multi-tree ts etc at the python level? We have a function get_example_tree_sequences() in test_highlevel.py which can return various pathological tree sequences like this. 

No, because we can't run valgrind when Python is running and would therefore not detect memory access issues (which are non-deterministic and differ across platforms, so you can't just say "it doesn't segfault, so it's fine").

There's loads of tests checking for empty tree sequences across the suite - the test_highlevel.py approach isn't how we're doing things any more.

**Comment by hyanwong** [comment]
> No, because we can't run valgrind when Python is running and would therefore not detect memory access issues (which are non-deterministic and differ across platforms, so you can't just say "it doesn't segfault, so it's fine").

Ah, OK. Apart from your suggestion above for all `parent[j] == -1`, I'm not familiar with the way you might check memory issues here. But now that I have implemented that, perhaps I can just add `test_multiroot_diff_iter` and `test_empty_diff_iter` functions to test_trees.c?

> There's loads of tests checking for empty tree sequences across the suite - the test_highlevel.py approach isn't how we're doing things any more.

Added anyway: I think there's no harm, and a function to produce a whole set of different semi-pathological tree sequences seems a useful thing to have around.

**Comment by jeromekelleher** [comment]
> Ah, OK. Apart from your suggestion above for all parent[j] == -1, I'm not familiar with the way you might check memory issues here. But now that I have implemented that, perhaps I can just add test_multiroot_diff_iter and test_empty_diff_iter functions to test_trees.c?

Once the tests are implemented valgrind is run on them automatically, checking for these errors.

File: python/_tskitmodule.c
Comment (benjeffery) [inline python/_tskitmodule.c:9878]: Shouldn't this be `options`?

File: c/tests/test_trees.c
Comment (benjeffery) [inline c/tests/test_trees.c:399]: For functions with options I've been using the pattern at https://github.com/tskit-dev/tskit/blob/master/c/tests/test_tables.c#L782 as I think it is clearer what's going on.

**Review by benjeffery — CHANGES_REQUESTED** [review]
Think I found am incorrect variable usage in the python C layer, and one comment about test style.

File: python/_tskitmodule.c
Comment (hyanwong) [reply python/_tskitmodule.c:9878]: Oh, very good catch. Thanks.

File: c/tests/test_trees.c
Comment (hyanwong) [reply c/tests/test_trees.c:399]: Sorry - is that the right link @benjeffery - I can't see anything about options there?

File: c/tests/testlib.c
Comment (jeromekelleher) [inline c/tests/testlib.c:51]: Fantastic! Thanks for putting these in, that's a major bonus.

Comment (hyanwong) [reply c/tests/testlib.c:51]: No probs. I just need to sort out the C linting (not set up on my system) and Ben's last comment, and this should be done, ready for 0.3.1.

File: c/tskit/trees.c
Comment (jeromekelleher) [reply c/tskit/trees.c:4511]: Nope, these look ideal, thanks (modulo issues @benjeffery points out)

File: c/tests/testlib.c
Comment (benjeffery) [reply c/tests/testlib.c:51]: I can c lint if needed.

File: c/tests/test_trees.c
Comment (benjeffery) [reply c/tests/test_trees.c:399]: Yeah, that's the right link - basically call the test function with each option. In the link:
```
    test_edge_table_with_options(0);
    test_edge_table_with_options(TSK_NO_METADATA);
```

Comment (hyanwong) [reply c/tests/test_trees.c:399]: Oh, I see. Sorry - I was being dense. I agree, much clearer.

File: c/tests/testlib.c
Comment (hyanwong) [reply c/tests/testlib.c:51]: Thanks @benjeffery - it would save me quite a lot of hassle if you could lint the C, so please do if that's OK. I've just pushed the commit with the "options" parameter, which is much better (thanks for the tip). I can also squash the commits into 1, or I guess you can, it it's easy (the first one, f9a4a972d346862c5ba22e4673b856a149256c01) is a separate PR, though, so perhaps leave that)

Comment (benjeffery) [reply c/tests/testlib.c:51]: Done, force pushed.

Comment (hyanwong) [reply c/tests/testlib.c:51]: Thanks a lot, Ben!

**Review by benjeffery — CHANGES_REQUESTED** [review]
Looking good, thanks @hyanwong! We just need a changelog before merge.

**Comment by hyanwong** [comment]
> Looking good, thanks @hyanwong! We just need a changelog before merge.

Done, thanks. Note that this includes the change in https://github.com/tskit-dev/tskit/pull/788 as the first commit in the PR, so that can be PR can be closed if this is merged.

**Comment by jeromekelleher** [comment]
Let's hold off on this until our bugfix releases have been shipped.

**Comment by hyanwong** [comment]
I think this is ready to merge @benjeffery ?

**Review by benjeffery — APPROVED** [review]
LGTM, lets merge.

**Review by jeromekelleher — APPROVED** [review]

**Comment by benjeffery** [comment]
@hyanwong I've just force pushed to resolve the conflict and add the change to the C changelog too as this is a change to both APIs.

**Pull request merged**
